## Machine Learning

### Laura Licerio

Considera la función: f(x,y) = 3x2 + 2xy + 15x + 6y2 - y, comenzando desde el punto inicial (x, y) = (-2, 3), encuentra el mínimo valor estimado para el cual esta función converge.

In [1]:
import numpy as np
import numpy.linalg as npl
import sympy
from sympy import *
import matplotlib.pyplot as plt
from matplotlib.gridspec import GridSpec

x, y = symbols("x y")

punto_inicio = np.vstack([-2,-3])
funcion = 3*x**2 + 2*x*y + 15*x + 6*y**2 - y
funcion

3*x**2 + 2*x*y + 15*x + 6*y**2 - y

In [2]:
f = lambdify('x,y',funcion)

In [3]:
#encontrar gradiente

f1x = funcion.diff(x)
f1xLam = lambdify('x,y',f1x)

f1y = funcion.diff(y)
f1yLam = lambdify('x,y',f1y)

f1 = [f1x,f1y]
gradiente = lambdify('x,y',f1)


In [4]:
#implementacion de algoritmo de gradiente descendiente

def gradientDescent(gradiente, punto_inicio, tamaño_paso = .085, precision = .01, max_iter = 3000, initError = 10):
    # configuracion inicial
    x,y = punto_inicio.astype(float)
    currIter, iterCoords = 0, [np.hstack([x,y])]
    error = initError
    currX = np.array([x,y])

    # inicio de las iteraciones
    
    while npl.norm(error) > precision and currIter < max_iter:
        currIter+=1 # se incrementa el contador
        prevX = currX.copy() # registra el ultimo valor de error
        
        # Ecuación del gradiente descendente
        currX -= tamaño_paso * np.array(gradiente(x,y)) # Nuevo X = (anterior) menos (tam_paso *  gradiente(anterior)).
        x,y = currX[0].copy(), currX[1].copy() # Update X.
        
        error = currX - prevX # error en base a cuanto nos movemos con el tamaño del paso y el gradiente
        iterCoords.append(np.hstack([x,y])) # registra X para graficarlo
        
    return currIter, currX, np.vstack(iterCoords)

pasos_totales, coordenada_final, coordenadas_intermedias = gradientDescent(gradiente,punto_inicio)

print(pasos_totales)
print(coordenada_final)
coordenadas_intermedias[:5]


9
[[-2.67007459]
 [ 0.52747521]]


array([[-2.        , -3.        ],
       [-1.745     ,  0.485     ],
       [-2.2125    ,  0.37195   ],
       [-2.4223565 ,  0.453686  ],
       [-2.53908131,  0.48772689]])

La coordenada mínima de f(x,y) es [[-2.67007459] [ 0.52747521]] , se necesitaron 9 pasos para alcanzarla

Se quiere vender cierto número N de artículos con el fin de maximizar la ganancia. La investigación de mercado dice que si se fija el precio en 1.50 , se podrá vender 5,000 artículos, y por cada 10 centavos que se reduzca el precio por debajo de 1.50 se podrá vender otros 1,000 artículos. 
Bajo el supuesto de que los costos fijos ("costos iniciales") totalizan 2,000, y el costo de producción por artículo ("costo marginal") es 0.50, encontrar el precio a establecer por artículo y la cantidad de artículos vendidos para maximizar las ganancias, y también determinar la ganancia máxima que puede obtener.


In [196]:
#por diferenciador de lagrange
p = symbols('p')

to_simpl = ( 5000 + 1000*(1.5 - p ) / 0.1 )* p - 2000 - 0.5* (5000 + 1000*(1.5 - p ) / 0.1 ) 


ganancia = (simplify(to_simpl))

print (ganancia)


-10000.0*p**2 + 25000.0*p - 12000.0


In [197]:
f_ganancia = lambdify('p', ganancia)

from sympy import *
l = symbols('l')
d_ganancia = ganancia.diff()


DL = np.array([
            d_ganancia 
        ])

In [198]:
solve([Eq(DL[0],0)],
       [p,l], simplify=False)



{p: 1.25000000000000}

In [199]:
ganancia.subs(p, 1.25)

3625.00000000000

In [200]:
n = 5000 + 1000*(1.5 - p ) / 0.1

In [201]:
n.subs(p,1.25)

7500.00000000000

La ganancia máxima es de 3625 estableciendo el precio a 1.25 y vendiendo 7500 articulos 